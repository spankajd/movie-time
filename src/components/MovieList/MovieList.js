import React, { useEffect, useState } from "react";
import MovieItem from "../MovieItem/MovieItem";
import './MovieList.scss';
import { useDispatch, useSelector } from 'react-redux';
import { selectSearchQuery, selectSelectedMovie, selectSortBy } from "../../store/selectors/movieSelectors";
import { selectMovie } from "../../store/actions/movieActions";


const MovieList = ({movies}) => {
    const dispatch = useDispatch();
    const [list, setList] = useState([]);
    const sortBy = useSelector(selectSortBy);
    const searchQuery = useSelector(selectSearchQuery);
    const selectedMovie = useSelector(selectSelectedMovie);

    useEffect( () => {
        let tempList = [...movies];

        if(searchQuery) {
            tempList = tempList.filter( movie => movie.title.includes(searchQuery));
            dispatch(selectMovie(tempList[0]));
        }

        if(sortBy) {
            tempList = tempList.sort((a, b) => {
                if (a[sortBy] === "N/A" && b[sortBy] === "N/A") return 0;
                if (!a[sortBy]) return 1;
                if (!b[sortBy]) return -1;
                return isNaN(a[sortBy]) || isNaN(b[sortBy]) ? a[sortBy].localeCompare(b[sortBy]) : parseFloat(a[sortBy]) - parseFloat(b[sortBy]);
              });
        }
        setList(tempList);

    },[movies, sortBy, searchQuery, dispatch]);

    return <div className="movie-list"> 
        {
            list?.map( movie => <MovieItem movie={movie} isActive={movie.title === selectedMovie?.title} key={movie.episode_id} />)
            }
    </div>
}

export default MovieList;