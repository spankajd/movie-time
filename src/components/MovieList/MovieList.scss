.movie-list {
    display: flex;
    flex-direction: column;
    padding: 0;
    margin: 0;
    list-style: none;

    width: 60%;


    .movie-item {
        border-bottom: 1px solid #8badce;
    }
    .movie-details {
        display: none;
    }

    @media (max-width: 768px) {
        .movie-details {
            display: flex !important;
            background: #FFF;
        }
    }
}