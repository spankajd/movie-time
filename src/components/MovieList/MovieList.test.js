import { screen } from "@testing-library/react";
import '@testing-library/jest-dom';
import { renderWithProviders } from "../../utils/test-utlis";
import MovieList from "./MovieList";

describe("MovieList Component", () => {
    const dummyMovies = [
        {
            title:"Sample Movie",
            episode_id:1,
            imdbRating: 'N/A'
        },
        {
            title:"Testing Movie",
            episode_id:2,
            imdbRating: 'N/A'
        },
        {
            title:"XYZ Movie",
            episode_id:3,
            imdbRating: '8.5'
        }
    ]

    it("renders MovieList", () => {
        renderWithProviders(
            <MovieList movies={[{title:'Test'}]}/>
        );
        const element = screen.getByText(/Test/i);
        expect(element).toBeInTheDocument();
    });
    
    it("matches snapshot", () => {
        const { asFragment } = renderWithProviders(<MovieList movies={[{title:'Test'}]} />);
        expect(asFragment()).toMatchSnapshot();
    });

    
    it("check with dummy search filter", () => {
        renderWithProviders(<MovieList movies={dummyMovies} />, {
            preloadedState: {
                moviesState: {
                    searchQuery: "Testing Movie",
                    movies:dummyMovies
                }
            }
        });
        const element = screen.getByText(/Episode 2/i);
        expect(element).toBeInTheDocument();
    });

    it("check with dummy sort by title filter", () => {
        renderWithProviders(<MovieList movies={dummyMovies} />, {
            preloadedState: {
                moviesState: {
                    sortBy: "title",
                    movies:dummyMovies
                }
            }
        });
        const element = screen.getByText(/Episode 1/i);
        expect(element).toBeInTheDocument();
    });

    it("check with dummy sort by episode id filter", () => {
        renderWithProviders(<MovieList movies={dummyMovies} />, {
            preloadedState: {
                moviesState: {
                    sortBy: "episode_id",
                    movies:dummyMovies
                }
            }
        });
        const element = screen.getByText(/Episode 1/i);
        expect(element).toBeInTheDocument();
    });

    it("check with dummy sort by rating filter", () => {
        renderWithProviders(<MovieList movies={dummyMovies} />, {
            preloadedState: {
                moviesState: {
                    sortBy: "imdbRating",
                    movies:dummyMovies
                }
            }
        });
        const element = screen.getByText(/Episode 1/i);
        expect(element).toBeInTheDocument();
    });

    it("check with dummy sort by missing property filter", () => {
        renderWithProviders(<MovieList movies={dummyMovies} />, {
            preloadedState: {
                moviesState: {
                    sortBy: "year",
                    movies:dummyMovies
                }
            }
        });
        const element = screen.getByText(/Episode 1/i);
        expect(element).toBeInTheDocument();
    });
});