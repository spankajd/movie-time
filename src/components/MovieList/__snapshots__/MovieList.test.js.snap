// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="movie-list"
  >
    <div
      class="movie-item false"
      data-testid="movie-item"
    >
      <div
        class="movie-id"
      >
        Episode 
      </div>
      <div
        class="movie-title"
      >
        Episode  - Test
      </div>
      <div
        class="movie-rating"
      >
        <div
          class="rating"
        >
          N/A
        </div>
      </div>
      <div
        class="movie-release-date"
      />
    </div>
  </div>
</DocumentFragment>
`;
